packages/block-editor/src/components/provider/index.js

```diff
@@ -15,6 +15,8 @@ import { compose } from '@wordpress/compose';
  */
 import withRegistryProvider from './with-registry-provider';
 
+/** @typedef {import('@wordpress/data').WPDataRegistry} WPDataRegistry */
```

Comment:
It's hard to tell.. are these actually resolving to anything? When I evaluate them in the editor, it just shows as the fallback "any" type.

![image](https://user-images.githubusercontent.com/1779930/69892633-7ba58300-12d5-11ea-99f1-515d9de6e661.png)

It's something where I think just like our modules have a public API for functions, etc, we'll want to see how to expose the types as part of the public interface. It seems like this can work by default when the `typedef` is defined in the entry point file of a module, but if it's not, we'll need to figure out a way to "re-export" it (I think it's enough to do the `typedef` aliasing in the entry point file).

I guess in any case, this helps avoid having the whitelist while still making the linter happy, then we can solve the problem of the public types definitions separately.

Reply:
I need to start using the proper IDE first to feel the experience 😂

How about we put all type definitions in the root index.js file of each package and import in all places where they are used? It should make it easier to manage.

Reply:
I see what you mean. The issue I see is that sometimes those imports are resolved with DefinitelyTyped, so I guess it should be tackled separately. Well, I opened Visual Studio Code for the first time since I don't remember 😅 

I tried two things:

- re-import the existing definition from the `index.js` file but it didn't work
- moving definitions to `index.js` and referencing them for `@wordpress/data` from there worked for all packages, but it didn't work for `@wordpress/element` when I used the same technique